pkg/ring/kv/memberlist/memberlist_client.go

```diff
@@ -335,49 +396,65 @@ func (m *KV) GetCodec(codecID string) codec.Codec {
 }
 
 // GetListeningPort returns port used for listening for memberlist communication. Useful when BindPort is set to 0.
+// This call is only valid after KV service has been started.
 func (m *KV) GetListeningPort() int {
 	return int(m.memberlist.LocalNode().Port)
 }
 
 // JoinMembers joins the cluster with given members.
 // See https://godoc.org/github.com/hashicorp/memberlist#Memberlist.Join
+// This call is only valid after KV service has been started and is still running.
 func (m *KV) JoinMembers(members []string) (int, error) {
+	if m.State() != services.Running {
+		return 0, fmt.Errorf("service not Running")
+	}
 	return m.memberlist.Join(members)
 }
 
-func (m *KV) retryJoinWithBackoff(members []string) {
+func (m *KV) joinMembersOnStartup(ctx context.Context, members []string) error {
+	reached, err := m.memberlist.Join(m.cfg.JoinMembers)
+	if err == nil {
+		level.Info(util.Logger).Log("msg", "joined memberlist cluster", "reached_nodes", reached)
+		return nil
+	}
+
+	if m.cfg.MaxJoinRetries <= 0 {
+		return err
+	}
+
+	level.Debug(util.Logger).Log("msg", "attempt to join memberlist cluster failed", "retries", 0, "err", err)
+	lastErr := err
+
 	cfg := util.BackoffConfig{
 		MinBackoff: m.cfg.MinJoinBackoff,
 		MaxBackoff: m.cfg.MaxJoinBackoff,
 		MaxRetries: m.cfg.MaxJoinRetries,
 	}
 
-	backoff := util.NewBackoff(context.Background(), cfg)
+	backoff := util.NewBackoff(ctx, cfg)
 
 	for backoff.Ongoing() {
+		backoff.Wait()
+
 		reached, err := m.memberlist.Join(members)
-		if reached == 0 {
-			level.Debug(util.Logger).Log("msg", "retry to join memberlist cluster", "node", m.cfg.NodeName, "attempts", backoff.NumRetries(), "err", err)
-			backoff.Wait()
+		if err != nil {
+			lastErr = err
+			level.Debug(util.Logger).Log("msg", "attempt to join memberlist cluster failed", "retries", backoff.NumRetries(), "err", err)
 			continue
 		}
 
-		level.Debug(util.Logger).Log("msg", "joined memberlist cluster", "node", m.cfg.NodeName, "reached", reached)
+		level.Info(util.Logger).Log("msg", "joined memberlist cluster", "reached_nodes", reached)
 		break
 	}
 
-	if backoff.Err() != nil {
-		level.Error(util.Logger).Log("msg", "failed to join memberlist cluster", "node", m.cfg.NodeName, "err", backoff.Err())
-	}
+	return lastErr
```

Comment:
I think there's a bug here: `lastErr` is not reset to `nil` before the `break` in the loop. I think we can simplify the logic switching `break` into `return nil`.

Reply:
Correct. Added test to catch this. Also added test for scenario when memberlist fails to join the cluster and abort flag is true.